src/metabase/util/queue.clj

```diff
@@ -63,24 +70,141 @@
   (DelayQueue.))
 
 (defn put-with-delay!
-  "Put an item on the delay queue, with a delay given in milliseconds."
+  "Put an item on a delay queue, with a delay given in milliseconds."
   [^DelayQueue queue delay-ms value]
   (.offer queue (->DelayValue value (.plus (Instant/now) (Duration/ofMillis delay-ms)))))
 
-(defn- take-delayed-batch* [^DelayQueue queue max-items ^long poll-ms acc]
+(defn- take-batch* [^BlockingQueue queue ^long max-messages ^long latest-time acc]
   (loop [acc acc]
-    (if (>= (count acc) max-items)
-      acc
-      (if-let [item (if (pos? poll-ms)
-                      (.poll queue poll-ms TimeUnit/MILLISECONDS)
-                      (.poll queue))]
-        (recur (conj acc (:value item)))
-        (not-empty acc)))))
-
-(defn take-delayed-batch!
-  "Get up to `max-items` of the ready items off a given delay queue."
-  ([queue max-items]
-   (take-delayed-batch* queue max-items 0 []))
-  ([^DelayQueue queue max-items ^long max-first-ms ^long max-next-ms]
-   (when-let [fst (.poll queue max-first-ms TimeUnit/MILLISECONDS)]
-     (take-delayed-batch* queue max-items max-next-ms [(:value fst)]))))
+    (let [remaining-ms (- latest-time (System/currentTimeMillis))]
+      (if (or (neg? remaining-ms) (>= (count acc) max-messages))
+        acc
+        (if-let [item (.poll queue remaining-ms TimeUnit/MILLISECONDS)]
+          (recur (conj acc
+                       (if (instance? DelayQueue queue)
+                         (:value item)
+                         item)))
+          (not-empty acc))))))
+
+(def ^:dynamic *take-batch-wait-ms*
+  "Time to wait for the first message to be available in take-batch! in milliseconds."
+  Long/MAX_VALUE)
+
+(defn take-batch!
+  "Get a batch of messages off the given queue.
+  Will wait for the first message to be available, then will collect up to max-messages in max-batch-ms, whichever comes first.
+  For convenience, if the queue is a DelayQueue, the returned values will be the actual values, not the Delay objects.
+
+  By default, will wait indefinitely for the first message, but that can be controlled with *take-batch-wait-ms*."
```

Comment:
I understand why a dynamic variable is useful for the listeners, as the only time we wouldn't want to block indefinitely is when testing, but I don't see any not to keep it a first class argument for this function.

Reply:
Yeah, I did it as a dynamic variable because I was thinking it shouldn't really be a part of the API but we need it for testing. Since it's actually only used in one place, I made it private and take the wait as a regular argument